reporter.go

```diff
@@ -0,0 +1,209 @@
+package k8sreporter
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NamespaceFilter is a filter function to choose what namespaces to dump.
+type NamespaceFilter func(string) bool
+
+// AddToScheme is a function for extend the reporter scheme and the CRs we are able to dump
+type AddToScheme func(*runtime.Scheme)
+
+// KubernetesReporter is a Ginkgo reporter that dumps info
+// about configured kubernetes objects.
+type KubernetesReporter struct {
+	sync.Mutex
+	clients        *clientSet
+	reportPath     string
+	namespaceToLog NamespaceFilter
+	crs            []CRData
+}
+
+// CRData represents a cr to dump
+type CRData struct {
+	Cr        runtimeclient.ObjectList
+	Namespace *string
+}
+
+// New returns a new Kubernetes reporter from the given configuration.
+func New(kubeconfig string, addToScheme AddToScheme, namespaceToLog NamespaceFilter, reportPath string, crs ...CRData) (*KubernetesReporter, error) {
+	crScheme := runtime.NewScheme()
+	clientgoscheme.AddToScheme(crScheme)
+	addToScheme(crScheme)
+
+	clients, err := newClient(kubeconfig, crScheme)
+
+	if err != nil {
+		return nil, err
+	}
+
+	crsToDump := []CRData{}
+	if crs != nil {
+		crsToDump = crs[:]
+	}
+
+	return &KubernetesReporter{
+		clients:        clients,
+		reportPath:     reportPath,
+		namespaceToLog: namespaceToLog,
+		crs:            crsToDump,
+	}, nil
+}
+
+// Dump dumps the relevant crs + pod logs
+func (r *KubernetesReporter) Dump(duration time.Duration, dirName string) {
+	since := time.Now().Add(-duration).Add(-5 * time.Second)
+
+	err := os.Mkdir(path.Join(r.reportPath, dirName), 0755)
+	if err != nil && !errors.Is(err, os.ErrExist) {
+		fmt.Fprintf(os.Stderr, "failed to create test dir: %v\n", err)
+		return
+	}
+	r.logNodes(dirName)
+	r.logLogs(since, dirName)
+	r.logPods(dirName)
+
+	for _, cr := range r.crs {
+		r.logCustomCR(cr.Cr, cr.Namespace, dirName)
+	}
+}
+
+func (r *KubernetesReporter) logPods(dirName string) {
+	pods, err := r.clients.Pods(v1.NamespaceAll).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to fetch pods: %v\n", err)
+		return
+	}
+	for _, pod := range pods.Items {
+		if !r.namespaceToLog(pod.Namespace) {
+			continue
+		}
+		f, err := logFileFor(r.reportPath, dirName, pod.Namespace+"-pods_specs")
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to open pods_specs file: %v\n", dirName)
+			return
+		}
+		defer f.Close()
+		fmt.Fprintf(f, "-----------------------------------\n")
```

Comment:
Should we have this delimiter as a const?

Reply:
done